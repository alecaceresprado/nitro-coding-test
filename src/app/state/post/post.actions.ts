import { createAction, props } from '@ngrx/store';

import { Post } from '@models';

export const fetchPosts = createAction('[POST] fetch posts');
export const fetchPostsSucceeded = createAction('[POST] fetch post - SUCCESS', props<{posts: Post[]}>());
export const fetchPostsFailed = createAction('[POST] fetch post - FAIL');
