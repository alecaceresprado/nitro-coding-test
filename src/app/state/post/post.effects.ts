import { Injectable } from '@angular/core';
import { Actions, createEffect, ofType } from '@ngrx/effects';
import { of } from 'rxjs';
import { switchMap, map, catchError } from 'rxjs/operators';

import { PostService } from '@services';
import {
  fetchPosts,
  fetchPostsSucceeded,
  fetchPostsFailed,
} from './post.actions';
import { Post } from '@models';

@Injectable()
export class PostEffects {
  /*
   * @listen [POST] fetch posts
   * @dispatch [POST] fetch post - SUCCESS
   * @dispatch [POST] fetch post - FAIL
   */
  public fetchPosts$ = createEffect(() =>
    this.actions$.pipe(
      ofType(fetchPosts),
      switchMap(() =>
        this.postService.getPosts().pipe(
          map((posts: Post[]) => fetchPostsSucceeded({ posts })),
          catchError(() => of(fetchPostsFailed()))
        )
      )
    )
  );

  constructor(private actions$: Actions, private postService: PostService) {}
}
