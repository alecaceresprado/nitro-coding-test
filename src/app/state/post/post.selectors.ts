import { createFeatureSelector } from '@ngrx/store';

import { Post } from '@models';

export const getPosts = createFeatureSelector<Post[]>(
  'posts'
);
