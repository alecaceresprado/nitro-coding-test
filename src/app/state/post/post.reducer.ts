import { createReducer, Action, on } from '@ngrx/store';

import { fetchPostsSucceeded, updatePost } from './post.actions';
import { Post } from '@models';

export const postInitialState: Post[] = [];

const reducer = createReducer(
  postInitialState,
  on(fetchPostsSucceeded, (state, { posts }) => posts),
  on(updatePost, (state, {id, author, location}) =>
    state.map(post => {
      if(post.id === id) {
        return {
          ...post,
          location,
          author
        };
      }
      return post;
    })
  )
);

export function postReducer(state: Post[] | undefined, action: Action): Post[] {
  return reducer(state, action);
}
