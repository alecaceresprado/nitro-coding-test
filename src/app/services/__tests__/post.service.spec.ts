import { of } from 'rxjs';

import { PostService } from '../post.service';
import { Post } from '@models';
const mockDb = require('../../../../db.json');

const httpClientMock = {
  get: jest.fn(() => of(mockDb.posts)),
};

describe('PostService', () => {
  let service: PostService;

  beforeAll(() => {
    service = new PostService(httpClientMock as any);
  });

  it('should get posts', () => {
    service.getPosts().subscribe((posts: Post[]) => {
      expect(posts).toMatchSnapshot();
    });
  });
});
