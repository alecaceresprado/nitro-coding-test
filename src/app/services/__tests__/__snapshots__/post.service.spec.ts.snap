// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`PostService should get posts 1`] = `
Array [
  Object {
    "author": "Happy User",
    "id": 1,
    "location": "San Francisco",
    "text": "Proper PDF conversion ensures that every element of your document remains just as you left it.",
    "time": "1552657573",
  },
  Object {
    "author": "Happy User",
    "id": 2,
    "location": "San Francisco",
    "text": "The modernworkplace is increasingly digital, and workflows are constantly evolving. ",
    "time": "1552571173",
  },
  Object {
    "author": "Happy Developer",
    "id": 3,
    "location": "SanFrancisco",
    "text": "Digital transformation isnâ€™t justa buzzword",
    "time": "1552571174",
  },
  Object {
    "author": "Happy Developer",
    "id": 4,
    "location": "Sydney",
    "text": "An expectation of digital efficiency has become the norm in our daily lives",
    "time": "1552563973",
  },
  Object {
    "author": "Happy Manager",
    "id": 5,
    "location": "Dublin",
    "text": "A modern PDF annotator that canaccommodate all of the cooks in a very busy kitchen is what your employees really need.",
    "time": "1553080742",
  },
  Object {
    "author": "Happy Manager",
    "id": 6,
    "location": "Dublin",
    "text": "An integratedproductivity solution breaks information through barriers and allows workers to collaborate in real time.",
    "time": "1553099742",
  },
]
`;
